### PR TITLE
[Merged by Bors] - feat(LinearAlgebra/Matrix/Permanent): smul

### PR DESCRIPTION
**Pull request by mo271**
adding 

-  `permanent_smul`
- `permanent_updateRow_smul`
- `permanent_updateColumn_smul`

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/Matrix/Permanent.lean [hunk 1]
```diff
@@ -83,4 +82,45 @@ theorem permanent_permute_rows (σ : Perm n) (M : Matrix n n R) :
...
+  congr
+  ext p
+  have h : p⁻¹ j ∈ univ := mem_univ _
+  repeat rw [← mul_prod_erase (Finset.univ ) _ h]
+  simp only [← mul_assoc, Perm.apply_inv_self, updateRow_self]
+  congr 1
+  apply Finset.prod_congr rfl
+  intro i hi
+  have : p i ≠ j := by aesop
+  simp [this]
```
Comment (loefflerd) [inline Mathlib/LinearAlgebra/Matrix/Permanent.lean:109]: If you do this after the columns version, you can do it in 2 lines:
```suggestion
@[simp]
theorem permanent_updateRow_smul (M : Matrix n n R) (j : n) (c : R) (u : n → R) :
    permanent (updateRow M j <| c • u) = c * permanent (updateRow M j u) := by
  rw [← permanent_transpose, ← updateColumn_transpose, permanent_updateColumn_smul,
    updateColumn_transpose, permanent_transpose]
```

File: Mathlib/LinearAlgebra/Matrix/Permanent.lean [hunk 2]
```diff
@@ -83,4 +82,45 @@ theorem permanent_permute_rows (σ : Perm n) (M : Matrix n n R) :
...
+  rw [Finset.mul_sum]
+  congr
+  ext p
+  have h : j ∈ univ := mem_univ _
+  repeat rw [← mul_prod_erase (Finset.univ ) _ h]
+  simp only [← mul_assoc, updateColumn_self]
+  congr 1
+  apply Finset.prod_congr rfl
+  intro i hi
+  simp [ne_of_mem_erase hi]
```
Comment (loefflerd) [inline Mathlib/LinearAlgebra/Matrix/Permanent.lean:124]: golfed version:
```suggestion
@[simp]
theorem permanent_updateColumn_smul (M : Matrix n n R) (j : n) (c : R) (u : n → R) :
    permanent (updateColumn M j <| c • u) = c * permanent (updateColumn M j u) := by
  simp only [permanent, ← mul_prod_erase _ _ (mem_univ j), updateColumn_self, Pi.smul_apply,
    smul_eq_mul, mul_sum, ← mul_assoc]
  congr 1 with p
  rw [Finset.prod_congr rfl (fun i hi ↦ ?_)]
  simp only [ne_eq, ne_of_mem_erase hi, not_false_eq_true, updateColumn_ne]
```
I fixed a typo `upodate` (you should probably fix this in the PR header). The `congr with` syntax is a handy shortcut for manipulating goals involving bound variables, it lets you dive inside the binder in a single step.

**Review by loefflerd — COMMENTED** [review]
Good stuff, thanks! I have a few shortcuts to suggest, nothing major.

**Comment by mo271** [comment]
Thanks for the review! Learned something...

**Comment by loefflerd** [comment]
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**